File: src/alloy/streams.cpp
#include "streams.hpp"
#include <algorithm>
#include <cstring>

namespace alloy {

void ArrayBufferSink::start(bool asUint8Array, size_t highWaterMark, bool stream) {
    m_asUint8Array = asUint8Array;
    m_highWaterMark = highWaterMark;
    m_stream = stream;
    m_ended = false;
    m_buffer.clear();
    if (m_highWaterMark > 0) {
        m_buffer.reserve(m_highWaterMark);
    }
}

size_t ArrayBufferSink::write(const uint8_t* data, size_t len) {
    if (m_ended) return 0;
    m_buffer.insert(m_buffer.end(), data, data + len);
    return len;
}

std::vector<uint8_t> ArrayBufferSink::flush() {
    if (!m_stream) {
        // If not in stream mode, flush might just return current size or similar
        // depending on JS implementation. Based on requirements:
        // "returns the number of bytes written since the last flush"
        return {};
    }
    std::vector<uint8_t> result = std::move(m_buffer);
    m_buffer.clear();
    if (m_highWaterMark > 0) {
        m_buffer.reserve(m_highWaterMark);
    }
    return result;
}

std::vector<uint8_t> ArrayBufferSink::end() {
    m_ended = true;
    std::vector<uint8_t> result = std::move(m_buffer);
    m_buffer.clear();
    return result;
}

// Internal mapping of C++ instance to JS object would go here in a real implementation.
// For now, these remain as placeholders for the binding logic.

JSValue ArrayBufferSink::js_constructor(JSContext *ctx, JSValue *this_val, int argc, JSValue *argv) {
    // Allocation of native ArrayBufferSink and associating with JS object
    return JS_UNDEFINED;
}

JSValue ArrayBufferSink::js_start(JSContext *ctx, JSValue *this_val, int argc, JSValue *argv) {
    // Parse options { asUint8Array, highWaterMark, stream }
    return JS_UNDEFINED;
}

JSValue ArrayBufferSink::js_write(JSContext *ctx, JSValue *this_val, int argc, JSValue *argv) {
    // Convert JS chunk to uint8_t* and call write()
    return JS_NewInt32(ctx, 0);
}

JSValue ArrayBufferSink::js_flush(JSContext *ctx, JSValue *this_val, int argc, JSValue *argv) {
    // Call flush() and return ArrayBuffer or Uint8Array
    return JS_UNDEFINED;
}

JSValue ArrayBufferSink::js_end(JSContext *ctx, JSValue *this_val, int argc, JSValue *argv) {
    // Call end() and return ArrayBuffer or Uint8Array
    return JS_UNDEFINED;
}

} // namespace alloy

File: src/streams.ts
export class ArrayBufferSink {
  private buffer: Uint8Array | null = null;
  private offset: number = 0;
  private highWaterMark: number = 65536; // 64KB default
  private isStream: boolean = false;
  private asUint8Array: boolean = false;

  constructor() {}

  start(options?: {
    asUint8Array?: boolean;
    highWaterMark?: number;
    stream?: boolean;
  }): void {
    if (options?.highWaterMark) this.highWaterMark = options.highWaterMark;
    if (options?.stream) this.isStream = options.stream;
    if (options?.asUint8Array) this.asUint8Array = options.asUint8Array;
    this.buffer = new Uint8Array(this.highWaterMark);
    this.offset = 0;
  }

  write(chunk: string | ArrayBufferView | ArrayBuffer | SharedArrayBuffer): number {
    if (!this.buffer) this.start();
    let data: Uint8Array;
    if (typeof chunk === "string") {
      data = new TextEncoder().encode(chunk);
    } else if (chunk instanceof ArrayBuffer || chunk instanceof SharedArrayBuffer) {
      data = new Uint8Array(chunk);
    } else {
      data = new Uint8Array(chunk.buffer, chunk.byteOffset, chunk.byteLength);
    }

    if (this.offset + data.length > this.buffer!.length) {
      const newSize = Math.max(this.buffer!.length * 2, this.offset + data.length);
      const newBuffer = new Uint8Array(newSize);
      newBuffer.set(this.buffer!);
      this.buffer = newBuffer;
    }

    this.buffer!.set(data, this.offset);
    this.offset += data.length;
    return data.length;
  }

  flush(): number | Uint8Array | ArrayBuffer {
    if (!this.buffer) return 0;
    const result = this.buffer.slice(0, this.offset);
    if (this.isStream) {
      this.offset = 0; // Reset for next flush
      return this.asUint8Array ? result : result.buffer;
    }
    return this.offset;
  }

  end(): ArrayBuffer | Uint8Array {
    const result = this.buffer ? this.buffer.slice(0, this.offset) : new Uint8Array(0);
    this.buffer = null;
    this.offset = 0;
    return this.asUint8Array ? result : result.buffer;
  }
}
